### ISSUE TITLE
tproxy method does not appear completely transparent to SSL over TCP

### ISSUE THREAD

**Issue by pisto** [sshuttle/sshuttle#371]
I tried using `sshuttle --method=tproxy` to route some Fortnite traffic out of a firewalled network. I found that UDP forwarding works fine, but TCP does not. Unfortunately I cannot debug this further as Fortnite is closed source. I'll just report my research into the issue.

It appears that somehow TCP traffic is not transparent with `sshuttle --method=tproxy` when the carried data is SSL. In the proxied application (Fortnite), it appears as if the TCP connections time out without a reply: for example, during the initial loading screen it does the "checking conection to datacenters", "looking for party to rejoin" etc but it times out for *most* of them, though unpredictably about which request and how often. From a wireshark dump I can see that the SSL connection is established, the reply is received by the application, but it must fail some validation at the SSL layer because it appears that the reply is simply discarded. The SSL connection is then left to timeout, and eventually it is gracefully closed, likely the result of some garbage collection within the application.

I tried to find out whether Fortnite is using some exotic TCP mechanism, but I couldn't pinpoint any. I am sure Fortnite does not use the TCP urgent pointer, which would not pass through sshuttle. So right now I'm stuck with using sshuttle just for the UDP part, but not TCP.

**Comment by brianmay** [comment]
I could be completely wrong, but but my guess is it is some sort of buffering deadlock situation. I have heard it can be tricky to get these details right - and took a lot of work to fix all the problems in openssh even. That is sshuttle has received data (meaning either sshuttle itself or the kernel) and isn't passing it on because it is expecting something else to happen (e.g. more data to fill the buffer), which doesn't happen, because the application is waiting for sshuttle to pass the data on.

Unfortunately I don't have the time or expertise to debug such problems.

I guess it is also possible that sshuttle is corrupting/dropping the data it is suppose to be passing through, I think this might be less likely however.

**Comment by pisto** [comment]
I don't believe it's a buffering issue. When I first found out about the problem I blamed sshuttle and made my own TPROXY tunnel (https://github.com/pisto/tfunnel), I got the same issue, and I did check that all data was sent and it was not sitting in a buffer. So I believe it's something intrinsic in the method.

**Comment by brianmay** [comment]
Interesting data point. It sounds like something wrong is going wrong in the kernel... What kernel version?

**Comment by pisto** [comment]
The TPROXY part runs on a 5.* machine and the remote part ran on a few different boxes with different kernels so I don't think the kernel version matter much, it was either 3.10 or 5.* as well anyway
